src/analyze_streams.cpp

```diff
@@ -0,0 +1,91 @@
+#include <migraphx/analyze_streams.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/errors.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+bool happens_before(const std::vector<std::size_t>& e1, const std::vector<std::size_t>& e2)
+{
+    return std::equal(e1.begin(), e1.end(), e2.begin(), e2.end(), std::less_equal<>{}) and
+           not std::equal(e1.begin(), e1.end(), e2.begin(), e2.end(), std::greater_equal<>{});
+}
+
+std::vector<stream_race> analyze_streams(const program& p, const stream_model& m)
+{
+    using vector_clock = std::vector<std::size_t>;
+    std::vector<stream_race> races;
+    auto nstream = m.get_nstream();
+    std::vector<vector_clock> vclock(nstream, vector_clock(nstream));
```

Comment:
can non-diagonal-elements be non-zero?

Reply:
Yea, otherwise it would be just a simple [Lamport timestamp](https://en.wikipedia.org/wiki/Lamport_timestamp).